locator/tablets.cc

```diff
@@ -245,9 +245,31 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
-    const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+future<tablet_map::primary_replica_map> tablet_map::calc_primary_replica_map(table_id table, const topology& topo, sstring dc) const {
+    std::unordered_map<host_id, uint64_t> load_map;
+    primary_replica_map res;
+    res.reserve(tablet_count());
+    co_await for_each_tablet([&] (tablet_id tid, const tablet_info& ti) {
+        tablet_replica_set replicas = dc.empty() ? ti.replicas :
+                boost::copy_range<tablet_replica_set>(ti.replicas | boost::adaptors::filtered([&] (const auto& tr) {
+                    const auto* node = topo.find_node(tr.host);
+                    return node && node->dc_rack().dc == dc;
+                }));
+        if (replicas.empty()) {
+            throw std::runtime_error(format("Could not find any replicas for table={} tablet={} dc={}", table, tid, dc));
+        }
+        std::ranges::sort(replicas, std::less{}, [&] (const tablet_replica& tr) -> uint64_t {
+            if (auto it = load_map.find(tr.host); it != load_map.end()) {
```

Comment:
Isn't `load_map` always empty at this point? I only see content being added below.

Reply:
Ah, I see this is in the `for_each_tablet()` loop, so it will have content by the second+ iterations.

Reply:
Maybe we can use a struct to hold the uint64_t value with default initialization to 0 and then
just return `load_map[tr.host]`

Reply:
used struct default initialization to simplify the sort cmp function v11